s3/pkg/datastore/yig/storage/multipart.go

```diff
@@ -1,53 +1,242 @@
 package storage
 
-/*import (
+import (
+	"context"
 	"crypto/md5"
-	"encoding/base64"
 	"encoding/hex"
 	"io"
-	"net/url"
 	"sort"
 	"strconv"
-	"time"
-
-	"github.com/opensds/multi-cloud/s3/pkg/datastore/yig/common"
-	"github.com/opensds/multi-cloud/s3/pkg/datastore/yig/crypto"
-	"github.com/opensds/multi-cloud/s3/pkg/datastore/yig/datatype"
-	. "github.com/opensds/multi-cloud/s3/pkg/datastore/yig/error"
-	"github.com/opensds/multi-cloud/s3/pkg/helper"
-	obj "github.com/opensds/multi-cloud/s3/pkg/datastore/yig/meta"
-	meta "github.com/opensds/multi-cloud/s3/pkg/datastore/yig/meta/types"
-	"github.com/opensds/multi-cloud/s3/pkg/datastore/yig/redis"
-)*/
-
-import (
-	"context"
-	"errors"
-	"io"
 
+	s3err "github.com/opensds/multi-cloud/s3/error"
+	"github.com/opensds/multi-cloud/s3/pkg/datastore/common"
+	"github.com/opensds/multi-cloud/s3/pkg/datastore/yig/meta/types"
 	"github.com/opensds/multi-cloud/s3/pkg/model"
 	pb "github.com/opensds/multi-cloud/s3/proto"
+	log "github.com/sirupsen/logrus"
 )
 
 const (
 	MAX_PART_SIZE   = 5 << 30 // 5GB
+	MIN_PART_SIZE   = 5 << 20 // 5MB
+	MIN_PART_NUMBER = 1
 	MAX_PART_NUMBER = 10000
 )
 
+/*
+* Below is the process of multipart upload:
+* 1. InitMultipartUpload will return a upload id and object id, and the caller should save them.
+* 2. caller should call UploadPart and input upload id and part number and then transfer the data.
+* backend will save the upload id and the part number and other related information.
+* 3. caller will call CompleteMultipartUpload and input the upload id, the backend
+* will mark all the parts identified by upload id as completed. If the caller calls
+* AbortMultipartUpload, backend will remove all the parts belongs to the upload id.
+*
+ */
+
 func (yig *YigStorage) InitMultipartUpload(ctx context.Context, object *pb.Object) (*pb.MultipartUpload, error) {
-	return nil, errors.New("not implemented.")
+	uploadId := uploadId2Str(yig.idGen.GetId())
+	mu := &pb.MultipartUpload{
+		Bucket:   object.BucketName,
+		Key:      object.ObjectKey,
+		UploadId: uploadId,
+		ObjectId: uploadId,
```

Comment:
Why use uploadId for ObjectId?

Reply:
because the final object consists of more than one parts, and when complete all the parts, yig backend doesn't really perform combine operation to reduce the pressure of io on the ceph. So, we use uploadId to track all the parts which belongs to the object.